pip/templates/deploy.py

```diff
@@ -30,48 +30,67 @@
-            '$DEPLOY_PIP_PASSWORD env variable'
-        )
-    command = ['./dist/*', '-u', pip_username, '-p', pip_password, '--repository-url', pip_registry]
+dist_dir = "./dist"
+PYPRC_KEY = 'pyprc'
+SNAPSHOT_KEY = 'snapshot'
+RELEASE_KEY = 'release'
+
+ENV_DEPLOY_PIP_USERNAME = 'DEPLOY_PIP_USERNAME'
+ENV_DEPLOY_PIP_PASSWORD = 'DEPLOY_PIP_PASSWORD'
+
+repositories = {
+    PYPRC_KEY: "{pyprc_repository}",
+    SNAPSHOT_KEY: "{snapshot}",
+    RELEASE_KEY: "{release}"
+}
+
+parser = argparse.ArgumentParser()
+parser.add_argument('repo_type')
+
+def upload_command(repo_type_key, package_file, wheel_file):
+    if repo_type_key not in repositories:
+        raise Exception(f"Selected repository must be one of: {list(repositories.keys())}")
+
+    if repo_type_key == PYPRC_KEY:
+        return [package_file, wheel_file, '--repository', repositories[repo_type_key]]
+    elif repo_type_key == SNAPSHOT_KEY or repo_type_key == RELEASE_KEY:
+        pip_username, pip_password = (os.getenv(ENV_DEPLOY_PIP_USERNAME), os.getenv(ENV_DEPLOY_PIP_PASSWORD))
+        if not pip_username:
+            raise Exception(f"username should be passed via the {ENV_DEPLOY_PIP_USERNAME} environment variable")
+        if not pip_password:
+            raise Exception(f"password should be passed via the {ENV_DEPLOY_PIP_PASSWORD} environment variable")
+        return [package_file, wheel_file, '-u', pip_username, '-p', pip_password, '--repository-url', repositories[repo_type_key]]
+    else:
+        raise Exception(f"Unrecognised repository selector: {repo_type_key}")
+
+
+if not os.path.exists("{package_file}"):
+    raise Exception("Cannot find expected distribution .tar.gz to deploy at '{package_file}'")
+
+if not os.path.exists("{wheel_file}"):
+    raise Exception("Cannot find expected distribution wheel to deploy at '{wheel_file}'")
+
+args = parser.parse_args()
+repo_type_key = args.repo_type
 
 with open("{version_file}") as version_file:
     version = version_file.read().strip()
-new_package_file = None
-new_wheel_file = None
 try:
-    dist_prefix = "./dist/"
-    if not os.path.exists(dist_prefix):
-        os.mkdir(dist_prefix)
-        
-    new_package_file = dist_prefix + "{package_file}".replace(".tar.gz", "-{}.tar.gz".format(version))
-    new_wheel_file = dist_prefix + "{wheel_file}".replace(".whl", "-{}.whl".format(version))
+    if not os.path.exists(dist_dir):
+        os.mkdir(dist_dir)
+
+    new_package_file = dist_dir + "/{package_file}".replace(".tar.gz", "-{}.tar.gz".format(version))
+    new_wheel_file = dist_dir + "/{wheel_file}".replace(".whl", "-{}.whl".format(version))
+
+    if not os.path.exists(os.path.dirname(new_package_file)):
+        os.makedirs(os.path.dirname(new_package_file))
 
-    if os.path.exists("{package_file}"):
-        shutil.copy("{package_file}", new_package_file)
+    if not os.path.exists(os.path.dirname(new_wheel_file)):
+        os.makedirs(os.path.dirname(new_wheel_file))
```

Comment:
Make paths that may not exist - the `{package_file}` contains its own sub-directories such as `grpc/python/python-protocol.tar.gz` that are not going to exist under `dist/` unless we create them